rosidl_generator_cpp/resource/msg__struct.hpp.em

```diff
@@ -81,55 +86,112 @@ struct @(spec.base_type.type)_
-  @('explicit ' if alloc_type else '')@(spec.base_type.type)_(@(alloc_type + alloc_name))
-@# generate initializer lists
-@[  if alloc_name == '_alloc']@
-// *INDENT-OFF* (prevent uncrustify from making unnecessary indents here)
-@[  end if]@
 @{
-used_alloc_name = False
-leading_colon = ':'
-for field in spec.fields:
-    # dynamic arrays require allocator if available
-    if alloc_name and field.type.is_array and \
-            (field.type.array_size is None or field.type.is_upper_bound):
-        if leading_colon == ' ':
-            print(',')
-        print('  %s %s(%s)' % (leading_colon, field.name, alloc_name), end='')
-        used_alloc_name = True
-        leading_colon = ' '
-
-    # if default value is available set it in initializer list
-    elif field.type.is_primitive_type() and field.default_value is not None:
-        cpp_value = value_to_cpp(field.type, field.default_value)
-        if leading_colon == ' ':
-            print(',')
-        print('  %s %s(%s)' % (leading_colon, field.name, cpp_value), end='')
-        leading_colon = ' '
-if leading_colon == ' ':
-    print()
+# The creation of the constructors for messages is a bit complicated.  The goal
+# is to have a constructor where the user can control how the fields of the
+# message get initialized via the _init parameter to the constructor.  See
+# http://design.ros2.org/articles/generated_interfaces_cpp.html#constructors
+# for a detailed explanation of the different _init parameters.
+init_list, alloc_list, member_list = create_init_alloc_and_member_lists(spec)
+
+def generate_default_string(membset):
+    strlist = []
+    for member in membset.members:
+        if member.default_value is not None:
+            if member.num_prealloc > 0:
+                strlist.append('this->%s.resize(%d);' % (member.name, member.num_prealloc))
+            if isinstance(member.default_value, list):
+                if member.default_value:
+                    if all(v == member.default_value[0] for v in member.default_value):
+                        strlist.append('std::fill(this->%s.begin(), this->%s.end(), %s);' % (member.name, member.name, member.default_value[0]))
+                    else:
+                        for index, val in enumerate(member.default_value):
+                            strlist.append('this->%s[%d] = %s;' % (member.name, index, val))
+            else:
+                strlist.append('this->%s = %s;' % (member.name, member.default_value))
+
+    return strlist
+
+def generate_zero_string(membset, fill_args):
+    strlist = []
+    for member in membset.members:
+        if isinstance(member.zero_value, list):
+            if member.num_prealloc > 0:
+                strlist.append('this->%s.resize(%d);' % (member.name, member.num_prealloc))
+            if member.zero_need_array_override:
+                strlist.append('this->%s.fill(%s {%s});' % (member.name, member.type, fill_args))
```

Comment:
Is there an example with this line expanded and being used? I just can't imagine how it is supposed to look like.

Reply:
If you run

`src/ament/ament_tools/scripts/ament.py test --build-tests --isolated --symlink-install --only rosidl_generator_cpp`

And then take a look at the `vec3_fixed` field in `./build_isolated/rosidl_generator_cpp/rosidl_generator_cpp/rosidl_generator_cpp/msg/primitives_some_default__struct.hpp`, that's an example that uses it.

Reply:
The line expands to:

```
this->vec3_fixed.fill(MyVector3 {_init});
```

Looks good to me.

Reply:
Looking at this a bit more: I have a question that I'm not sure to be able to answer just based on the design doc for interfaces.
Should I be able to use a message from the same package in a service definition without specifying the package it comes from?
Example because I know the previous sentence is not clear:
I have a package as follow:
```
├── CMakeLists.txt
├── msg
│   └── foo.msg
├── package.xml
└── srv
    └── bar.srv
```
Does bar.srv **have to** be:
```
pkgA/foo[] fools
---
pkgA/foo[3] threefools
```
or is the following a valid definition?
```
foo[] fools
---
foo[3] threefools
```

Reading the [design doc](http://design.ros2.org/articles/interface_definition.html):
```
The base type can be one of the following:

- a primitive type from the list above: e.g. int32
- a string with an upper boundary: string<=N to limit the length of the string to N characters
- a complex type referencing another message:
- an absolute reference of a message: e.g. some_package/SomeMessage
- a relative reference of a message within the same package: e.g. OtherMessage
```
 it seems that the answer should be "both are valid". 

If both are valid, we need to update this part of the PR (and maybe others?) to use the full c++ namespace of the message and not only the type name.
`this->vec3_fixed.fill(MyVector3 {_init});` -> `this->vec3_fixed.fill(rosidl_generator_cpp::msg::MyVector3 {_init});`

If only the definition using "absolute references" is allowed, we should update the design document to make it clearer.

Example branch that fails to compile with the current patch: https://github.com/ros2/rcl_interfaces/tree/more_services

Reply:
This is an excellent point.  Based on the design document, I agree with your reasoning, and relative should be allowed.  I'll take a closer look at your branch and see what I have to do to fix it here.  Thanks for the heads up.